packages/@aws-cdk/core/lib/private/refs.ts

```diff
@@ -199,8 +201,13 @@ function getCreateExportsScope(stack: Stack) {
 }
 
 function generateExportName(stackExports: Construct, id: string) {
+  const stackRelativeExports = stackExports.construct.tryGetContext(cxapi.STACK_RELATIVE_EXPORTS_CONTEXT);
   const stack = Stack.of(stackExports);
-  const components = [...stackExports.construct.scopes.slice(2).map(c => c.construct.id), id];
+
+  const components = stackRelativeExports
+    ? [...stackExports.construct.scopes.slice(stack.construct.scopes.length).map(c => c.construct.id), id]
+    : [...stackExports.construct.scopes.slice(2).map(c => c.construct.id), id];
+
```

Comment:
I feel this makes it easier to notice where the difference introduced by the feature flag is:

```suggestion
  const components = [
    ...stackExports.construct.scopes
      .slice(stackRelativeExports ? stack.construct.scopes.length : 2)
      .map(c => c.construct.id),
    id
  ];

```